### PR TITLE
To and create.

### PR DESCRIPTION
**Pull request by AGaliciaMartinez**
This PR adds:
- DenseTensor class that wraps around a TensorFlow `Tensor`.
- `to` and `create` functions for the dispatcher.

### CODE REVIEW

**Comment by AGaliciaMartinez** [comment]
One of the things I had to do here is set copy=False in the __init__ of `DenseTensor`. This is so that `create` maintains the graph structure as I would say this should be the default behaviour. This is different to what happens with `Dense` where the default value of `copy` is `True`. Should I keep it false to maintain the graph structure or is there a way to keep the graph by default when using create but still having `copy=True` in `DenseTensor`?

File: src/qutip_tensorflow/core/data/convert.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+# You can add user friendly name for conversion with `to` or Qobj creation functions:
+qt.data.to.register_aliases(['tftensor'], DenseTensor)
+qt.data.to.register_aliases(['DenseTensor'], DenseTensor)
+
+qt.data.create.add_creators([
+    (is_tftensor, DenseTensor, 85),
+])
+
+to = qt.data.to
+create = qt.data.create
```
Comment (AGaliciaMartinez) [inline src/qutip_tensorflow/core/data/convert.py:32]: Should I make to and create accessible from qtf?

File: src/qutip_tensorflow/core/data/convert.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
+import qutip as qt
```
Comment (jakelishman) [inline src/qutip_tensorflow/core/data/convert.py:1]: I would argue there's no reason to abbreviate `qutip`.  It's short already, and it just hurts readability.

File: src/qutip_tensorflow/core/data/convert.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+# `add_conversions` will register the data layer
+qt.data.to.add_conversions([
+     (DenseTensor, qt.data.Dense, _tf_from_dense),
+     (qt.data.Dense, DenseTensor, _tf_to_dense),
+])
+
+# You can add user friendly name for conversion with `to` or Qobj creation functions:
+qt.data.to.register_aliases(['tftensor'], DenseTensor)
+qt.data.to.register_aliases(['DenseTensor'], DenseTensor)
```
Comment (jakelishman) [inline src/qutip_tensorflow/core/data/convert.py:25]: the first argument is a list, specifically so you can pass more than one thing in a single call

File: src/qutip_tensorflow/core/data/convert.py (hunk 1 above)
Comment (jakelishman) [reply src/qutip_tensorflow/core/data/convert.py:32]: No - there's almost never a reason to re-expose the same object from a different location, particularly not across library boundaries.  The point of a dispatcher plugin is that the data-types should work from the main QuTiP library for basic use-cases, otherwise you'll cause confusion - if a user mutates the `qutip_tensorflow.to` object (e.g. by adding aliases or changing the preferred conversion rules), it would be a huge surprise to discover that `qutip.to` has also changed.

File: src/qutip_tensorflow/core/data/dense_tensor.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
+import qutip as qt
```
Comment (jakelishman) [inline src/qutip_tensorflow/core/data/dense_tensor.py:1]: again, really not sure I see a reason to abbreviate this.  People read code more than they write it.

File: src/qutip_tensorflow/core/data/dense_tensor.py [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+import numpy as np
+import tensorflow as tf
+import qutip.core.data as data
```
Comment (jakelishman) [inline src/qutip_tensorflow/core/data/dense_tensor.py:4]: You should protect this name with an underscore or something - you frequently use `data` as a variable, so at several points during the code, it's not at all clear what the identifier `data` refers to.

File: src/qutip_tensorflow/core/data/dense_tensor.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+    """
+
+    def __init__(self, data, shape=None, copy=False):
+        # Try to inherit shape from data
+        if shape is None:
+            try:
+                shape = data.shape
+            except AttributeError:
+                raise ValueError("""Shape could not be inferred from data.
+                                 Please, include the shape of data.""")
```
Comment (jakelishman) [inline src/qutip_tensorflow/core/data/dense_tensor.py:35]: It's not clear to me what inputs you're prepared to accept.  The docstrings says `array_like`, but this check will cause `DenseTensor([[1, 0], [0, 1]])` to raise an exception, despite it being a valid Numpy array-like object.  You might want to rethink the order you do things, so you're not repeating checks or throwing away valid input from which the shape _can_ be inferred.  Remember that the `__init__` method is really meant for _humans_ more than anything else - library code often uses fast alternative constructor methods, because it often already has a lot of knowledge about the data being passed in, and having to write more doesn't matter very much in library code.

File: src/qutip_tensorflow/core/data/dense_tensor.py [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+        ):
+            raise ValueError("""shape must be a 2-tuple of positive ints, but
+                             is """ + repr(shape))
+
+        super().__init__(shape)
+
+        try:
+            self._tf = tf.constant(data, shape=shape)
+        except TypeError as e:
+            raise ValueError("Shape of data must match shape argument.") from e
```
Comment (jakelishman) [inline src/qutip_tensorflow/core/data/dense_tensor.py:62]: This error handling doesn't look right.  What if `data` is some random object which has a `shape` attribute, but isn't array-like?  Then the error is that invalid data has been passed, but `tf.constant(data)` will still likely throw a TypeError, so you'll see this error.

File: tests/core/data/test_tensordense.py [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+
+    def test_copy_returns_a_correct_copy(self, data_tensor_dense):
+        """
+        Test that the copy() method produces an actual copy, and that the
+        result represents the same matrix.
+        """
+        original = data_tensor_dense
+        copy = data_tensor_dense.copy()
+        assert original is not copy
+        assert np.all(original._tf == copy._tf)
```
Comment (jakelishman) [inline tests/core/data/test_tensordense.py:223]: You should also check that mutations of the original don't affect the copy, and vice-versa.

File: tests/core/data/test_tensordense.py [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+        # by default we do not return a copy
+        assert test._tf is tensor_dense
+
+    @pytest.mark.parametrize('dtype', ['complex128',
+                                       'float64',
+                                       'int32', 'int64',
+                                       'uint32'])
+    def test_init_from_list_other_dtype(self, shape, dtype):
+        numpy_dense = np.random.rand(*shape).astype(dtype, casting='unsafe')
+        list_dense = numpy_dense.tolist()
```
Comment (jakelishman) [inline tests/core/data/test_tensordense.py:85]: You're shadowing a fixture name here, which is unclear - it means the same name refers to quite different things through the code.

File: tests/core/data/test_tensordense.py [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+        test = DenseTensor(list_dense, numpy_dense.shape)
+        assert test.shape == shape
+        assert test._tf.dtype == tf.complex128
+        assert test._tf.shape == shape
+        assert_almost_equal(test.to_array(), list_dense)
```
Comment (jakelishman) [inline tests/core/data/test_tensordense.py:90]: Are they really allowed to only be _almost_ equal?

File: tests/core/data/test_tensordense.py [hunk 11]
```diff
@@ -0,0 +1,234 @@
...
+    def test_init_from_wrong_input(self, arg, kwargs, error):
+        """
+        Test that the __init__ method raises a suitable error when passed
+        incorrectly formatted inputs.
+
+        This test also serves as a *partial* check that Dense safely handles
+        deallocation in the presence of exceptions in its __init__ method.  If
+        the tests segfault, it's quite likely that the memory management isn't
+        being done correctly in the hand-off us setting our data buffers up and
+        marking the numpy actually owns the data.
```
Comment (jakelishman) [inline tests/core/data/test_tensordense.py:146]: Please don't just copy-and-paste code - incorrect documentation can even be worse than no documentation at all.

File: src/qutip_tensorflow/core/data/dense_tensor.py (hunk 7 above)
Comment (AGaliciaMartinez) [reply src/qutip_tensorflow/core/data/dense_tensor.py]: I changed code logic. I think now this issue would be solved.

File: tests/core/data/test_tensordense.py (hunk 8 above)
Comment (AGaliciaMartinez) [reply tests/core/data/test_tensordense.py:223]: What do you mean by mutations? Could you provide me an example with when this could go wrong?  

If by mutations you mean that I should check that an "in place" modification of original._tf does no change copy._tf (for example by doing something like: `original._tf[0, 0] += 1`) then, it turns out that for `Tensor` it is not possible to modify it in place. For instance, `original._tf[0, 0] += 1` will return `TypeError`. `Variables` from TensorFlow can but I plan to implement them in another class.

In any case, thinking about your comment I realised that this can happen:
```python
original = data_tensor_dense
maybe_copy = DenseTensor(original._tf, copy=False) # This is not what happens with copy() method but could happen!
print(original is not maybe_copy) # True
print(np.all(original._tf == maybe_copy._tf)) #True
print(maybe_copy._tf is not original._tf) # False!!!!!!
```
So I will add in the tests: `assert copy._tf is not original._tf`. In any case I am not sure if this would matter that much as tensors can not be modified "in place".

File: tests/core/data/test_tensordense.py (hunk 9 above)
Comment (AGaliciaMartinez) [reply tests/core/data/test_tensordense.py:85]: Thanks for catching this. I named them _list_dense _numpy_dense. I would suggest doing the same in QuTiP tests as numpy_dense is sometimes used as a fixture and some other times not. ( I actually thought it did not matter because QuTiP tests already did it). I can make a pull request if in QuTiP if you agree.

File: tests/core/data/test_tensordense.py (hunk 10 above)
Comment (AGaliciaMartinez) [reply tests/core/data/test_tensordense.py]: With `assert np.all(test.to_array()==np.array(_list_dense))` it fails for:

FAILED tests/core/data/test_tensordense.py::TestClassMethods::test_init_from_list_other_dtype[ket-float64] - assert False
FAILED tests/core/data/test_tensordense.py::TestClassMethods::test_init_from_list_other_dtype[bra-float64] - assert False
FAILED tests/core/data/test_tensordense.py::TestClassMethods::test_init_from_list_other_dtype[square-float64] - assert False
FAILED tests/core/data/test_tensordense.py::TestClassMethods::test_init_from_list_other_dtype[wide-float64] - assert False
FAILED tests/core/data/test_tensordense.py::TestClassMethods::test_init_from_list_other_dtype[tall-float64] - assert False


Could this be due to the casting rules? I looks weird as it only fails for float64.

File: tests/core/data/test_tensordense.py (hunk 11 above)
Comment (AGaliciaMartinez) [reply tests/core/data/test_tensordense.py:146]: Sorry for this. I will go through all the comments in the test and ensure they are correct.

File: tests/core/data/test_tensordense.py [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+        pytest.param((4, 2), id="tall"),
+    ]
+)
+def shape(request):
+    return request.param
+
+
+@pytest.fixture(params=[True, False], ids=["Fortran", "C"])
+def fortran(request):
+    return request.param
```
Comment (AGaliciaMartinez) [inline tests/core/data/test_tensordense.py:31]: Remove this

File: tests/core/data/test_tensordense.py [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+        [
+            pytest.param(1, id="shape_no_tuple"),
+            pytest.param((), id="shape_0_tuple"),
+            pytest.param((1,), id="shape_1_tuple"),
+            pytest.param((None, None), id="shape_None_tuple"),
+            pytest.param([2, 2], id="shape_list"),
+            pytest.param((1, 2, 3), id="shape_3_tuple"),
+            pytest.param((-1, 1), id="negative_shape"),
+            pytest.param((-4, -4), id="both_negative_shape"),
+            pytest.param((1213, 1217), id="differen_shape"),
```
Comment (AGaliciaMartinez) [inline tests/core/data/test_tensordense.py:141]: check what happens if shape input not the same as data input

File: tests/core/data/test_tensordense.py (hunk 12 above)
Comment (AGaliciaMartinez) [reply tests/core/data/test_tensordense.py:31]: Actually, do not remove this as `numpy_dense` creates an array that depends on this fixture. For example, it necessary to test that `DenseTensor` is created properly from a Fortran array. Although, I am not sure how important is this.

**Comment by AGaliciaMartinez** [comment]
I added the missing test and renamed DenseTensor to TfTensor. I would like to merge this PR by tomorrow (@Ericgig agrees). @jakelishman, would you like to suggest any other change or are you happy with the merge?

**Review by Ericgig — APPROVED** [review]

File: src/qutip_tensorflow/core/data/tftensor.py [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+    shape: (int, int)
+        Shape of data. Default `None`, it tries to infer the shape from data
+        accessing the attribute `data.shape`.
+    copy: bool
+        If `True` (default) then the object is copied. Otherwise, copy will only
+        be made if tf.constant returns a copy of data(when input is not a
+        `Tenor`) or if a copy is needed to satisfy dtype (tensorflow.complex128) and
+        shape (2D Tensor)."""
+
+    def __init__(self, data, shape=None, copy=False):
```
Comment (jakelishman) [inline src/qutip_tensorflow/core/data/tftensor.py:27]: Is there any situation in which the backing data field of this object can be mutable?  I don't know tensorflow so well, but say for numpy
```python
>>> class Data:
...   def __init__(self, data):
...      self.data = np.array(data, copy=False)
...   def __repr__(self):
...     return f"Data({repr(self.data)})"
...
>>> x = np.ones((2,))
>>> y = Data(x)
>>> x
array([1., 1.])
>>> y
Data(array([1., 1.]))
>>> x *= 2
>>> x
array([2., 2.])
>>> y
Data(array([2., 2.]))
```
so the derived object has changed from under us.

If there is such a case, then the default argument to `copy` _must_ be `True`.  If not, then does the `copy` keyword argument serve any purpose?

File: src/qutip_tensorflow/core/data/tftensor.py [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+from tensorflow.errors import InvalidArgumentError
+import qutip
+import numbers
+
+__all__ = ["TfTensor"]
+
+
+class TfTensor(qutip.core.data.Data):
+    """This class provide a wraps around TensorFlow's Tensor. It will store data
+    as a Tensor of dtype `tensorflow.complex128`. Data will be expanded into a 2D
```
Comment (jakelishman) [inline src/qutip_tensorflow/core/data/tftensor.py:11]: you need double back-ticks (everywhere) to make this render as inline code, as opposed to in italics.

File: src/qutip_tensorflow/core/data/tftensor.py [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+
+        # Try to inherit shape from data and expand shape
+        if shape is None:
+            try:
+                shape = data.shape
+            except AttributeError:
+                raise ValueError(
+                    """Shape could not be inferred from data.
+                                 Please, include the shape of data."""
+                )
```
Comment (jakelishman) [inline src/qutip_tensorflow/core/data/tftensor.py:44]: `data` is now guaranteed to be a tensor, so is it ever possible for this to raise an exception?

File: src/qutip_tensorflow/core/data/tftensor.py [hunk 17]
```diff
@@ -0,0 +1,98 @@
...
+
+            if isinstance(shape, tf.TensorShape):
+                shape = tuple(shape.as_list())
```
Comment (jakelishman) [inline src/qutip_tensorflow/core/data/tftensor.py:47]: Similarly, is it ever possible for this _not_ to be the case?

File: src/qutip_tensorflow/__init__.py [hunk 18]
```diff
@@ -1 +1,2 @@
 from .version import version as __version__
+from .core import *
```
Comment (jakelishman) [inline src/qutip_tensorflow/__init__.py:2]: Do you need a separate `core` package within `qutip_tensorflow`?  I wouldn't imagine there'll be too many different levels in it, so this may just be adding needless structure.

File: tests/core/data/test_convert.py [hunk 19]
```diff
@@ -0,0 +1,80 @@
...
+@pytest.mark.parametrize(
+    ["to_", "dtype"],
+    [
+        pytest.param("tftensor", TfTensor, id="to tensorflow str"),
+        pytest.param("TfTensor", TfTensor, id="to tensorflow str_type"),
+        pytest.param(TfTensor, TfTensor, id="to tensorflow type"),
+    ],
+)
+def test_converters_qt_to_qtf(from_, base, to_, dtype):
+    print(base)
```
Comment (jakelishman) [inline tests/core/data/test_convert.py:75]: this `print` doesn't serve a purpose - pytest eats the stdout stream on success, and would print the parameters to the test on failure.

File: tests/core/data/test_convert.py [hunk 20]
```diff
@@ -0,0 +1,80 @@
...
+        pytest.param("TfTensor", TfTensor, id="to tensorflow str_type"),
+        pytest.param(TfTensor, TfTensor, id="to tensorflow type"),
+    ],
+)
+def test_converters_qtf_to_qt(from_, base, to_, dtype):
+    converter = data.to[to_, from_]
+    assert isinstance(converter(base), dtype)
+    converter = data.to[to_]
+    assert isinstance(converter(base), dtype)
+    assert isinstance(data.to(to_, base), dtype)
```
Comment (jakelishman) [inline tests/core/data/test_convert.py:52]: I'm not sure what exactly the test (and the similar reverse test) is meant for.  The behaviour seems to be a lot more of a test of `qutip.data.to` than anything else - it seems to just be testing that the alias and partial application systems work.  If your intent is to test that the various conversions have truly been added, you probably want to investigate the `direct` attribute of the output of `data.to[to_, from_]`.

File: tests/core/data/test_tftensor.py [hunk 21]
```diff
@@ -0,0 +1,171 @@
...
+        assert test.shape == numpy_dense.shape
+        assert np.all(test.to_array() == numpy_dense)
+
+    def test_init_from_tensor(self, tensor_dense):
+        test = TfTensor(tensor_dense)
+        assert test.shape == tuple(tensor_dense.shape.as_list())
+        assert np.all(test.to_array() == tensor_dense)
+
+        # by default we do not return a copy
+        assert test._tf is tensor_dense
```
Comment (jakelishman) [inline tests/core/data/test_tftensor.py:86]: If so, you probably also want to check that using the `copy` keyword _does_ return a copy.

File: tests/core/data/test_tftensor.py [hunk 22]
```diff
@@ -0,0 +1,171 @@
...
+        assert test.shape == shape
+        assert test._tf.dtype == tf.complex128
+        assert test._tf.shape == shape
+        assert_almost_equal(test.to_array(), _list_dense)
+
+    @pytest.mark.parametrize(
+        "dtype", ["complex128", "float64", "int32", "int64", "uint32"]
+    )
+    def test_init_from_ndarray_other_dtype(self, shape, dtype):
+        _numpy_dense = np.random.rand(*shape).astype(dtype, casting="unsafe")
```
Comment (jakelishman) [inline tests/core/data/test_tftensor.py:104]: Be careful with this casting; `np.random.rand()` returns a float on [0, 1], which will be cast to a integer 0 every time bar ~once every 2^53 times.  So you're basically just testing on the zero matrix every time, which probably isn't what you meant.

**Review by jakelishman — COMMENTED** [review]
I have some comments, a few of which may have been addressed in meetings that I've not been in (sorry).  Mostly it looks fine, though, with Eric's comment too, provided Eric's also happy with the default `copy=False` in the constructor.

File: src/qutip_tensorflow/__init__.py (hunk 18 above)
Comment (AGaliciaMartinez) [reply src/qutip_tensorflow/__init__.py:2]: I was following the same structure that QuTiP has for consistency. It is no necessary but I though someone working on QuTiP would find it more appealing.

File: src/qutip_tensorflow/core/data/tftensor.py (hunk 14 above)
Comment (AGaliciaMartinez) [reply src/qutip_tensorflow/core/data/tftensor.py:27]: The above case would not work for tensorflow if using a `Tensor`:
```python
a = random_tensor((1,1)) # random tensor of shape (1,1)
b = TfTensor(a)
b._tf == a # True
a *= 2
b._tf == a #False
```
To my knowledge, there is no operation in a `tf.Tensor` that does not return a different tensor. `tf.Variables`, although not supported yet in qutip-tensorflow, do perform in-place operations with the `assign()` method. 

I can not think of a case where `copy=true` is necessary but it does have a different behaviour. For instance,
```python
a = tf.constant([[1+1j]])
b = TfTensor(a, copy=False) 
a is not b._tf # False
b = TfTensor(a, copy=True)
a is not b._tf # True 
``` 
Although I do not know when and why this could be useful (this may be why `tf.constant()` does not have a `copy` argument).

File: src/qutip_tensorflow/core/data/tftensor.py (hunk 16 above)
Comment (AGaliciaMartinez) [reply src/qutip_tensorflow/core/data/tftensor.py:44]: You are right, this was necessary for the previous logic but is not any more. Removing.

File: tests/core/data/test_tftensor.py (hunk 22 above)
Comment (AGaliciaMartinez) [reply tests/core/data/test_tftensor.py:106]: I see the problem but given there is such a low chance to occur I think I will let it be for the moment and focus on the next PR.

File: tests/core/data/test_convert.py (hunk 20 above)
Comment (AGaliciaMartinez) [reply tests/core/data/test_convert.py:52]: This is quite a change and I really do not want to delay this any more. I will merge this pull request so that I can open the next. I will also open a new PR making this test more sensible.

File: src/qutip_tensorflow/core/data/tftensor.py (hunk 14 above)
Comment (jakelishman) [reply src/qutip_tensorflow/core/data/tftensor.py:27]: If tensors are immutable, then the copy argument is pretty redundant. It's fine to have it for consistency, you may just want to mention that it does very little in the docstring.

**Comment by AGaliciaMartinez** [comment]
Thanks @jakelishman and @Ericgig for reviewing the PR! I will merge this now and fix some minor comments in another PR. This way the next PR will be ready for review today.

File: src/qutip_tensorflow/__init__.py (hunk 18 above)
Comment (jakelishman) [reply src/qutip_tensorflow/__init__.py:2]: I'd say there's no need to try and maintain structural consistency between separate packages. QuTiP's core is just for code organisation anyway - it's not directly exposed to users.

File: tests/core/data/test_tftensor.py (hunk 22 above)
Comment (jakelishman) [reply tests/core/data/test_tftensor.py:106]: The point is more that the test isn't correct - you're not actually passing it random data, you're just passing a zero matrix. You could basically just fix it by generating random unsigned integers instead of random floats, and then you wouldn't need the unsafe cast either.
